server/services/v1/auth_provider.go

```diff
@@ -0,0 +1,357 @@
+	}
+	return &api.DeleteApplicationResponse{
+		Deleted: true,
+	}, nil
+}
+
+func (a *Auth0) UpdateApplication(ctx context.Context, req *api.UpdateApplicationRequest) (*api.UpdateApplicationResponse, error) {
+	// validate
+	client, err := a.Management.Client.Read(req.GetId())
+	if err != nil {
+		return nil, api.Errorf(a.managementToTigrisErrorCode(err), "Failed to update application: reason = %s", err.Error())
+	}
+
+	// check ownership before updating
+	currentSub, err := getCurrentSub(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to update application: reason = %s", err.Error())
+	}
+	if client.GetClientMetadata()[createdBy] != currentSub {
+		return nil, api.Errorf(api.Code_PERMISSION_DENIED, "Failed to update application: reason = You cannot update application that is not created by you.")
+	}
+
+	metadata := client.GetClientMetadata()
+	metadata[updatedBy] = currentSub
+	metadata[updatedAt] = time.Now().Format(time.RFC3339)
+	appToUpdate := &management.Client{
+		Name:           &req.Name,
+		Description:    &req.Description,
+		ClientMetadata: metadata,
+	}
+
+	err = a.Management.Client.Update(req.GetId(), appToUpdate)
+	if err != nil {
+		log.Warn().Msgf("Failed to update app clientId=%s, reason=%s", req.GetId(), err.Error())
+		return nil, api.Errorf(a.managementToTigrisErrorCode(err), "Failed to update application: reason = %s", err.Error())
+	}
+
+	client, err = a.Management.Client.Read(req.GetId())
+	if err != nil {
+		return nil, api.Errorf(a.managementToTigrisErrorCode(err), "Failed to update application: reason = %s", err.Error())
+	}
+	return &api.UpdateApplicationResponse{
+		UpdatedApplication: &api.Application{
+			Id:          client.GetClientID(),
+			Name:        client.GetName(),
+			Description: client.GetDescription(),
+			Secret:      client.GetClientSecret(),
+			CreatedAt:   readDate(client.GetClientMetadata()[createdAt]),
+			CreatedBy:   client.GetClientMetadata()[createdBy],
+			UpdatedAt:   readDate(client.GetClientMetadata()[updatedAt]),
+			UpdatedBy:   client.GetClientMetadata()[updatedBy],
+		},
+	}, nil
+}
+
+func (a *Auth0) RotateApplicationSecret(ctx context.Context, req *api.RotateApplicationSecretRequest) (*api.RotateApplicationSecretResponse, error) {
+	// validate
+	client, err := a.Management.Client.Read(req.GetId())
+	if err != nil {
+		return nil, api.Errorf(a.managementToTigrisErrorCode(err), "Failed to rotate application secret: reason = %s", err.Error())
+	}
+
+	// check ownership before rotating
+	currentSub, err := getCurrentSub(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to delete application: reason = %s", err.Error())
+	}
+	if client.GetClientMetadata()[createdBy] != currentSub {
+		return nil, api.Errorf(api.Code_PERMISSION_DENIED, "Failed to rotate application secret: reason = You cannot rotate secret for application that is not created by you.")
+	}
```

Comment:
The code from the beginning of the function up to this point is identical in three above functions can we dedup? (1)

Reply:
refactored.